std/typecons.d

```diff
@@ -2094,6 +2100,343 @@ enum isTuple(T) = __traits(compiles,
+    }
+    enum string hw = ctfeFunction(); // evaluated at compile-time
+    static assert(hw == "Hello, World!");
+}
+
+/// `tie` supports moving. It can be observed using an explicit destructor.
+pure nothrow @nogc @safe
+unittest
+{
+    static struct Move
+    {
+        int payload; // Some payload.
+        // Destructor to observe moving as
+        // the object will be reset to Move.init.
+        // Otherwise, the struct is too trivial
+        // and moving equals copying.
+        ~this() { }
+        // Check the value.
+        bool opEquals(int value) { return payload == value; }
+        // Check for reset.
+        bool isInit() { return payload == 0; }
+    }
+    // You can cache the result of tie[ ... ]
+    // but rvalues in right-hand side tie are moved out.
+    Move x = Move(5), y;
+    // Create the Tie
+    auto t = tie[ x, Move(2) ]; // Copies x, moves Move(2).
+
+    // Extract from the Tie #1.
+    tie[ y, $ ] = t;
+    assert(x == 5); // x was copied, not moved, to t[0] when creating t
+    assert(t[0] == 5); // t[0] was copied ...
+    assert(y == 5);  // ... to y.
+    assert(t[1] == 2); // t[1] untouched
+
+    y = Move(1);
+
+    // Extract from the Tie #2.
+    tie[ y, x ] = t;
+    assert(y == 5);  // t[0] was copied to y again.
+    assert(t[1].isInit); // t[1] moved ...
+    assert(x == 2); // ... to x
+
+    // Extract from the Tie manually.
+    import std.algorithm.mutation : move;
+    x = move(t[0]); // Move copied value manually.
+    assert(x == 5 && t[0].isInit);
+
+    // While the expression tie[ ... ] can be assigned,
+    // cached/returned results of tie[ ... ] cannot be assigned.
+    // Note that tie does not contain references and
+    // the expression
+    //     tie[ ... ] = source();
+    // does not create a Tie object; it calls opIndex(Op)Assign.
+    static assert(!__traits(compiles,
+    {
+        t = tie[ x, Move(2) ];
+    }));
+    static assert(!__traits(compiles,
+    {
+        t = t;
+    }));
+    // This is a feature. Tie does not use internal pointers,
+    // and is therefore inherently memory safe.
+
+    // Moving and static arrays.
+    assert(y == 5);
+    tie[ x ] = cast(Move[1]) [ move(y) ]; // rvalue (cast necessary)
+    assert(x == 5);
+    Move[2] marray = [ Move(3), Move(4) ];
+    tie[ x, $ ] = move(marray); // moved lvalue is being copied (see Limitations)
+    assert(x == 3);
+    // assert(marray[0].isInit); // Supposed to hold (see Limitations) ...
+    assert(marray[0] == 3); // ... instead of this.
+    // Warning: You should not rely on this behavior.
+
+    // Moving and Tuples.
+    auto mtuple = tuple(Move(2), Move(3));
+    tie[ x, $ ] = move(mtuple); // moved lvalue ...
+    assert(x == 2);
+    assert(mtuple[0].isInit); // ... from here
+    assert(mtuple[1].isInit); // destroyed
+    // Tuples do not support piecewise move extraction
+}
+
+/// `tie` supports non copyable and must-initialize types.
+pure nothrow @nogc @safe
+unittest
+{
```

Comment:
Again, a long unittest.

Reply:
Yes, in contrast to the other, is one is concerned about must-initialize and non-copyable types. Splitting it means duplicating the struct. The unit test is long, splitting it makes it longer. There's nothing to be left out. If you're dealing with non-copyable types, there's enough struggle. At least, the examples here will be rich.